feat: Finish Datadog account sync job

VDT1805 opened the pull request:
This PR introduces a new job queue system using `solid_queue` and includes various related changes to integrate it into the application. Additionally, it adds a new job for syncing Datadog accounts and corresponding tests.

### Integration of `solid_queue`:

* **Gemfile**: Added `solid_queue` gem dependency.
* **Configuration**:
  * [`config/environments/development.rb`](diffhunk://#diff-d3c4b3f41072daa416f1920511e9b2e26caea8c5cec0a14cb9508589a4dafa47R82-R84): Configured `solid_queue` as the queue adapter for Active Job.
  * [`config/environments/production.rb`](diffhunk://#diff-da60b4e96eff2b132991226d308949e23f4ef3aad45ad59edd09cbc32cc6251eL74-R76): Configured `solid_queue` as the queue adapter for Active Job.
  * [`config/environments/test.rb`](diffhunk://#diff-77f322f5ab0c54b1f2793b339574333dc55889645d910a93ede0fd7aa13b217aR68-R70): Configured `solid_queue` as the queue adapter for Active Job.
  * [`config/puma.rb`](diffhunk://#diff-24409379bdb75ad446bb4e2c18fd4cced1a263b99a3ff96fc3777d8fd8faeab9R35-R37): Added `solid_queue` plugin.
  * [`config/queue.yml`](diffhunk://#diff-4d10d2b1fe9107f3caf7333627fc8a59513845642bfdb2f5665635e7490f7307R1-R18): Added configuration for `solid_queue`.

### Database Changes:

* **Schema**:
  * [`db/queue_schema.rb`](diffhunk://#diff-f397d00b3e6805b27316dc41045504e9a356cb145e7bf8d601640b6682e26d4dR1-R129): Defined schema for `solid_queue` related tables.
* **Database Configuration**:
  * [`config/database.yml`](diffhunk://#diff-5a674c769541a71f2471a45c0e9dde911b4455344e3131bddc5a363701ba6325R22-R40): Added queue databases for development, test, and production environments. [[1]](diffhunk://#diff-5a674c769541a71f2471a45c0e9dde911b4455344e3131bddc5a363701ba6325R22-R40) [[2]](diffhunk://#diff-5a674c769541a71f2471a45c0e9dde911b4455344e3131bddc5a363701ba6325R63-R71)

### New Job for Datadog Sync:

* **Job Implementation**:
  * [`app/jobs/datadog_sync_job.rb`](diffhunk://#diff-63ad836d79d4eab311e5a6d388d15dfbd00e3c3fa86a1e24005dd9396d559d27R1-R36): Created `DatadogSyncJob` to sync Datadog accounts.
* **Model Hook**:
  * [`app/models/datadog.rb`](diffhunk://#diff-fe59497e60bf0e9d426b41c72f4e3b9e7fe22194fd749ef1e8045376847004f6R12-R14): Added after save commit hook to trigger `DatadogSyncJob`.

### Testing:

* **VCR Configuration**:
  * [`spec/cassettes/DatadogSyncJob/Datadog_Sync_Job_Behavior/when_fail/Datadog_accounts_sync_unsuccessful.yml`](diffhunk://#diff-195a67b16c10dd4a040344160a517122148531bf681ba399a266f00b27647c05R1-R41): Added VCR cassette for unsuccessful Datadog sync.
* **RSpec**:
  * [`spec/jobs/datadog_sync_job_spec.rb`](diffhunk://#diff-b212eb8ad026b535a94033975472994b3eaa96c8dd150dc660d8dac628e3d58cR1-R45): Added tests for `DatadogSyncJob`.
  * [`spec/spec_helper.rb`](diffhunk://#diff-89eebfcbc0f14b6d989517837ca1e94fce4e2ce9a03233641cd936f2b8d2ed94L28-R31): Updated sensitive data filtering for Datadog keys.

These changes collectively introduce `solid_queue` as the new job queue system and add functionality to sync Datadog accounts, along with necessary configurations and tests.

xuantien-raksul commented on app/jobs/datadog_sync_job.rb:
```diff
@@ -0,0 +1,36 @@
+      if response.status == 200
+            data = JSON.parse(response.body)["data"]
+
+            # Transform data for Account model
```
Let's remove all code comments. If you need to explain your code, give it a good name (function name, variable names, etc...)

xuantien-raksul commented on app/jobs/datadog_sync_job.rb:
```diff
@@ -0,0 +1,36 @@
+      conn = Faraday.new(
+        url: "https://api.datadoghq.com",
+      )
+      response = conn.get do |req|
+        req.url "/api/v2/users"
+        req.headers["Content-Type"] = "application/json"
+        req.headers["DD_API_KEY"] = job[:api_key]
+        req.headers["DD_APPLICATION_KEY"] = job[:application_key] # HAS TO BE DD_APPLICATION_KEY
+        req.headers["DD_SITE"] = job[:subdomain]
+      end
```
We're getting data via 1 API call only. Can you handle the case where we cannot fetch all data in a single API? (Pagination)

xuantien-raksul commented on app/jobs/demo_job.rb:
Please remove this

xuantien-raksul commented on app/models/datadog.rb:
```diff
@@ -6,7 +6,10 @@ class Datadog < ApplicationRecord
   validates_with DatadogValidator
 
+  after_save_commit do
+    DatadogSyncJob.perform_now({ api_key: api_key, application_key: application_key, subdomain: subdomain, connection_id: self.cred.connection_id })
```
Should it be perform_async? If you use perform_now, does it run the job in the main thread? Let's re-confirm this

xuantien-raksul replied:
Also you do not need to pass too much data into our job. We just need an ID, then within the job, we can query for api_key, application_key, subdomain, etc... from the data base

VDT1805 commented:
Thank you for the input. I have overhauled the code based on your input
### Job and Queuing System Enhancements:
* Added `BatchDatadogSyncJob` and `DatadogSyncJob` to handle Datadog API synchronization tasks (`app/jobs/batch_datadog_sync_job.rb`, `app/jobs/datadog_sync_job.rb`). [[1]](diffhunk://#diff-a5ee2e7cdcef2c38a1f741759d81f41dbae1e455e96fe6e22743522d7c9a96bfR1-R31) [[2]](diffhunk://#diff-63ad836d79d4eab311e5a6d388d15dfbd00e3c3fa86a1e24005dd9396d559d27R1-R30)
### Datadog Integration:
* Triggered `BatchDatadogSyncJob` after saving a `Datadog` record (`app/models/datadog.rb`).
* Added VCR cassettes for testing Datadog API interactions (`spec/cassettes/DatadogSyncJob/Datadog_Sync_Job_Behavior/when_fail/Datadog_accounts_sync_unsuccessful.yml`, `spec/cassettes/DatadogSyncJob/Datadog_Sync_Job_Behavior/when_success/Datadog_accounts_sync_successful.yml`). [[1]](diffhunk://#diff-195a67b16c10dd4a040344160a517122148531bf681ba399a266f00b27647c05R1-R41) [[2]](diffhunk://#diff-fa319a0f9ad0b0f98d9c307dadb74c3bb62fb12c795f40fff8c237396962ddadR1-R52)

### RuboCop Configuration:
* Updated `.rubocop.yml` to enforce specific layout styles (`.rubocop.yml`).
